common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -338,6 +340,79 @@ jbyteArray CBBToByteArray(JNIEnv* env, CBB* cbb) {
     return byteArray.release();
 }
 
+jbyteArray CryptoBufferToByteArray(JNIEnv* env, const CRYPTO_BUFFER* buf) {
+  if (CRYPTO_BUFFER_len(buf) > INT_MAX) {
+      JNI_TRACE("buffer too large");
+      Errors::jniThrowRuntimeException(env, "buffer too large");
+      return nullptr;
+  }
+
+  int length = static_cast<int>(CRYPTO_BUFFER_len(buf));
+  jbyteArray ret = env->NewByteArray(length);
+  if (ret == nullptr) {
+    JNI_TRACE("allocating byte array failed");
+    return nullptr;
+  }
+
+  env->SetByteArrayRegion(ret, 0, length, reinterpret_cast<const int8_t*>(CRYPTO_BUFFER_data(buf)));
+  return ret;
+}
+
+bssl::UniquePtr<CRYPTO_BUFFER> ByteArrayToCryptoBuffer(JNIEnv* env, const jbyteArray array,
+                                                       CRYPTO_BUFFER_POOL* pool) {
+    if (array == nullptr) {
+        JNI_TRACE("array was null");
+        Errors::jniThrowNullPointerException(env, "array == null");
+        return nullptr;
+    }
+
+    ScopedByteArrayRO arrayRo(env, array);
+    if (arrayRo.get() == nullptr) {
+        JNI_TRACE("failed to get bytes");
+        return nullptr;
+    }
+
+    bssl::UniquePtr<CRYPTO_BUFFER> ret(CRYPTO_BUFFER_new(
+            reinterpret_cast<const uint8_t*>(arrayRo.get()), arrayRo.size(), nullptr));
+    if (!ret) {
+        JNI_TRACE("failed to allocate CRYPTO_BUFFER");
+        Errors::jniThrowOutOfMemory(env, "failed to allocate CRYPTO_BUFFER");
+        return nullptr;
+    }
+
+    return ret;
+}
+
+static jobjectArray CryptoBuffersToObjectArray(JNIEnv* env,
+                                               const STACK_OF(CRYPTO_BUFFER)* buffers) {
+    size_t numBuffers = sk_CRYPTO_BUFFER_num(buffers);
+    if (numBuffers > INT_MAX) {
+        JNI_TRACE("too many buffers");
+        Errors::jniThrowRuntimeException(env, "too many buffers");
+        return nullptr;
+    }
+
+    ScopedLocalRef<jobjectArray> array(
+            env, env->NewObjectArray(static_cast<int>(numBuffers), JniConstants::byteArrayClass,
+                                     nullptr));
+    if (array.get() == nullptr) {
+        JNI_TRACE("failed to allocate array");
+        return nullptr;
+    }
+
+    for (size_t i = 0; i < numBuffers; ++i) {
+        CRYPTO_BUFFER* buffer = sk_CRYPTO_BUFFER_value(buffers, i);
+        ScopedLocalRef<jbyteArray> bArray(env, CryptoBufferToByteArray(env, buffer));
+        if (bArray.get() == nullptr) {
+            return nullptr;
+        }
+        env->SetObjectArrayElement(array.get(), i, bArray.get());
```

Comment:
Are you sure? I assumed it took another reference to the object but that it didn't consume the local reference or anything. The original code was used get() instead of release(), as does this code in Chromium (I just searched for the first hit for SetObjectArrayElement). Since JNI normally expects you not to free local references at all and just hope you not exceed 255, I expect taking ownership isn't really a thing JNI does.

https://cs.chromium.org/chromium/src/base/android/jni_array.cc?rcl=1d3d2786fbc2ca8c2fd371dde1c6f6449e7c1b43&l=121